### PR TITLE
npm utilities improvement

### PR DESCRIPTION
**Pull request by NotExpectedYet**
Couldn't use those packages they both relied on importing npm, which is broken it seems due to a filter command. 

Made enhancements to the original to make it detect the missing packages and install individually.

### CODE REVIEW

File: server_src/lib/serverCommands.js [hunk 1]
```diff
@@ -131,7 +131,12 @@ class SystemCommands {
           "<b class='text-danger'>Cancel:</b> This option will cancel the update process and not install the required dependencies. No update will run and manual intervention by the user is required. <br><br>";
         return clientResponse;
       } else {
-        await installMissingNpmDependencies();
+        if (missingPackagesList.length > 0) {
+          for (let i = 0; i > missingPackagesList.length; i++) {
+            await installNpmDependencies(missingPackagesList[i]);
+          }
+        }
+
```
Comment (NotExpectedYet) [inline server_src/lib/serverCommands.js:139]: Just changed here to cycle through the dependency and fire any missing packages to the updated function

File: server_src/utils/npm.utils.js [hunk 2]
```diff
@@ -1,37 +1,39 @@
 const util = require("util");
 const exec = util.promisify(require("child_process").exec);
 
-async function doWeHaveMissingPackages() {
-  const { stdout, stderr } = await exec("npm outdated --production");
+async function returnListOfMissingPackages() {
+  try {
+    const outdatedPackageJsonList = await exec(
+      "npm outdated --production --json"
+    );
```
Comment (NotExpectedYet) [inline server_src/utils/npm.utils.js:8]: Made npm outdated parsable by using it's --json function.

File: server_src/utils/npm.utils.js [hunk 3]
```diff
@@ -1,37 +1,39 @@
...
-    if (!!missingPackages) {
-      return missingPackages.length > 0;
-    } else {
-      return false;
+      keys.forEach((key, index) => {
+        if (!outdatedPackageListParsed[key].current) {
+          missingPackageList.push(key);
+        }
+      });
+      return missingPackageList;
```
Comment (NotExpectedYet) [inline server_src/utils/npm.utils.js:21]: After grabbing the missing list keys and cycling through them I check for the "current" key missing on the returned list and push the key which is the name of the package to the array.

File: server_src/utils/npm.utils.js [hunk 4]
```diff
@@ -1,37 +1,39 @@
...
-    throw `Could not execute npm install, is npm installed? | ${stderr}`;
-  }
-
-  if (stdout) {
-    return true;
+async function installNpmDependencies(dependency) {
+  try {
+    await exec(`npm install ${dependency}`);
+  } catch (e) {
+    throw `Error running installation command | ${e}`;
```
Comment (NotExpectedYet) [inline server_src/utils/npm.utils.js:32]: Not too much different just got rid of the stdout / stderr define and left it with a try/catch similar to the rest of the update functions.

File: server_src/lib/serverCommands.js [hunk 5]
```diff
@@ -4,8 +4,8 @@ const { isNodemon } = require("../utils/env.utils.js");
 const { lookpath } = require("lookpath");
 
 const {
-  doWeHaveMissingPackages,
-  installMissingNpmDependencies,
+  returnListOfMissingPackages,
+  installNpmDependencies,
```
Comment (NotExpectedYet) [inline server_src/lib/serverCommands.js:8]: These names are more understandable no? xD

Best be... *shakes fist*

File: test/application/server-commands.test.js [hunk 6]
```diff
@@ -114,7 +114,7 @@ describe("ServerCommands", () => {
     beforeEach(() => {
       mockedSimpleGit.setTestScenario(scenarioUpToDate);
       mockedSimpleGit.setIsRepo(true);
-      npmUtils.setHasMissingPackages(false);
+      npmUtils.setHasMissingPackages([]);
```
Comment (NotExpectedYet) [inline test/application/server-commands.test.js:117]: Array empty now means no packages

File: test/application/server-commands.test.js [hunk 7]
```diff
@@ -141,7 +141,7 @@ describe("ServerCommands", () => {
 
     it("should be able to detect but not fix missing npm packages without force.doWeInstallPackages", async () => {
       mockedSimpleGit.setTestScenario(scenarioBehindOutput);
-      npmUtils.setHasMissingPackages(true);
+      npmUtils.setHasMissingPackages(["random"]);
```
Comment (NotExpectedYet) [inline test/application/server-commands.test.js:144]: array package name we have missing deps

File: server_src/lib/serverCommands.js [hunk 8]
```diff
@@ -131,7 +130,9 @@ class SystemCommands {
           "<b class='text-danger'>Cancel:</b> This option will cancel the update process and not install the required dependencies. No update will run and manual intervention by the user is required. <br><br>";
         return clientResponse;
       } else {
-        await installMissingNpmDependencies();
+        for (let i = 0; i > missingPackagesList.length; i++) {
+          await installNpmDependencies(missingPackagesList[i]);
+        }
```
Comment (NotExpectedYet) [inline server_src/lib/serverCommands.js:135]: I use a for loop here due to the async await

Comment (NotExpectedYet) [reply server_src/lib/serverCommands.js:135]: It's checked initially to make sure it's got a length worth iterating over. If not it's the previous return user status anyway.

File: server_src/lib/serverCommands.js (hunk 5 above)
Comment (davidzwa) [reply server_src/lib/serverCommands.js:8]: I can save this line of code and comment and show it to you half a year later? 🦆

Comment (NotExpectedYet) [reply server_src/lib/serverCommands.js:8]: !RemindMe 1 year

File: server_src/lib/serverCommands.js [hunk 9]
```diff
@@ -119,10 +119,9 @@ class SystemCommands {
     await pullLatestRepository(force.forcePull);
 
     // Check to see if npm packages are missing and if so install them...
-    const missingPackages = await doWeHaveMissingPackages();
-
+    const missingPackagesList = await returnListOfMissingPackages();
     // If we have missing packages alert the user and wait for their response, if response given then install missing deps.
-    if (missingPackages) {
+    if (missingPackagesList.length > 0) {
```
Comment (davidzwa) [inline server_src/lib/serverCommands.js:124]: Falsy check?

Comment (NotExpectedYet) [reply server_src/lib/serverCommands.js:124]: That's always going to return an array, empty or not.

File: server_src/lib/serverCommands.js (hunk 8 above)
Comment (davidzwa) [reply server_src/lib/serverCommands.js:135]: `installNpmDependencies` should not be plural

File: server_src/utils/npm.utils.js (hunk 3 above)
Comment (davidzwa) [reply server_src/utils/npm.utils.js:21]: Object.entries() not better readable/understandable?

Comment (NotExpectedYet) [reply server_src/utils/npm.utils.js:21]: I didn't want a key value pair, only the key really.

**Review by davidzwa — APPROVED** [review]
Didnt test it but looks good

**Pull request merged**